### ISSUE TITLE
port ArmVirtPkg to the core PciHostBridgeDxe (Bugzilla Bug 65)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#6590]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 65
Date: 2016-07-20T21:23:09+00:00
From: @mdkinney
To:   @ardbiesheuvel
CC:   @lersek

Depends on:   #66
See also:     https://github.com/tianocore/edk2/issues/84
Last updated: 2016-09-02T17:16:37+00:00

**Comment by tianocore-issues** [comment]
## Comment 90
Date: 2016-07-20 21:23:09 +0000
From: @mdkinney

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): @ardbiesheuvel

rsek commented on May 5 • edited
The main customization in ArmVirtPkg/PciHostBridgeDxe is that it emulates IO port accesses with an MMIO range, whose base address is parsed from the DTB.

The core PciHostBridgeDxe driver delegates the IO port access implementation to EFI_CPU_IO2_PROTOCOL. @ardbiesheuvel recently implemented ArmPkg/Drivers/ArmPciCpuIo2Dxe (in commit 3dece14) which provides this protocol, backed by the same kind of translation as described above. The base address comes from gArmTokenSpaceGuid.PcdPciIoTranslation.

Therefore:

We should extend our ArmVirtPkg/Library/FdtPciPcdProducerLib plugin library to locate the IO translation offset in the DTB, and store it in the above PCD.
We should include ArmPciCpuIo2Dxe in the ArmVirtQemu builds, plugging the above library into it.
We should implement a PciHostBridgeLib instance, and plug it into the core PciHostBridgeDxe driver.
We should create one PCI_ROOT_BRIDGE object, populating it with the FTD Client code we are currently using in ArmVirtPkg/PciHostBridgeDxe.
Extra benefit: the core PciHostBridgeDxe driver supports 64-bit MMIO, we just have to parse those values from the DTB as well.
 @lerseklersek added Feature ArmVirtPkg labels on May 5
@lersek
     tianocore member
lersek commented on May 6
After briefly reviewing

gitk -- ArmPlatformPkg/ArmVirtualizationPkg/PciHostBridgeDxe
gitk -- ArmVirtPkg/PciHostBridgeDxe
I think the only real hurdle is PcdKludgeMapPciMmioAsCached (commit f9a8be4). I have no idea how to keep that hack with the core host bridge driver. :(

Maybe we&apos;ll postpone this until after we have a virtio-gpu driver, which should make VGA and PcdKludgeMapPciMmioAsCached unnecessary.
 @ardbiesheuvel
     tianocore member
ardbiesheuvel commented on May 6
To be honest, I am not as convinced as I was before that KVM is doing the right thing here.
This has nothing to do with DMA being coherent or not, since the BARs don&apos;t describe memory. So I now think that it is a bug for KVM to back anything that it exposes via PCI MMIO BARs with cacheable memory without keeping it synchronized between the guest and the host.
 @lersek
     tianocore member
lersek commented on May 6
Well, if it can be fixed in KVM, I won&apos;t object! :)
 @lersek
     tianocore member
lersek commented on May 25
The priority of this issue is increasing. 64-bit MMIO is really useful for virtio-pci devices (the modern ones, i.e., with disable-legacy=on,disable-modern=off). I wonder if we could keep PcdKludgeMapPciMmioAsCached in our new PciHostBridgeLib somehow.
 @lersek
     tianocore member
lersek commented on May 25
(BTW, I retested VGA on my Mustang just now, with PcdKludgeMapPciMmioAsCached=FALSE, and yes, it does break.)
 @ardbiesheuvel
     tianocore member
ardbiesheuvel commented on Jun 16
It is expected that the kludge PCD is still needed, unfortunately.

I don&apos;t think updating the core PciHostBridgeDxe driver to accommodate cached mappings for PCI regions is not the way to go. We really need to fix this in KVM instead.
 @lersek
     tianocore member
lersek commented on Jun 17
I recently hazarded to call this issue an &quot;ARMv8 architecture bug&quot; internally, expecting to be beaten down with arguments. To my surprise, I saw someone confirm my words, stating that the ARMv8 designers must not have been thinking with their &quot;virt hats on&quot; (I&apos;m paraphrasing).

If that&apos;s the case, can we really expect KVM to work around an architectural bug? Shouldn&apos;t we request new silicon / chipset releases from hw vendors instead, and/or an architectural update from ARM?

I mean, can a solution be conceived at all for KVM, with acceptable performance?
 @ardbiesheuvel
     tianocore member
ardbiesheuvel commented on Jun 18
Generally speaking, you could call it an oversight that the stage 2 mappings can only supersede guest stage 1 mappings in a way that makes them more strict. I think I may have heard a justification at some point that sounded plausible, but I am not able to reproduce it.

However, that does not mean that KVM is not broken. The fact that KVM has a cached mapping of a memory region that it can reasonably expect to be mapped uncached by the guest already violates the (given) architecture, although you could argue that those mappings are not live at the same time. But having KVM use uncached mappings for such regions would address both issues.
 @lersek
     tianocore member
lersek commented on Jun 18
Okay. Can you champion this on kvm-arm? I wouldn&apos;t know where to begin. :)
 @lersek
     tianocore member
lersek commented on Jun 18
(I don&apos;t mean patches -- I mean convincing others to write patches. :))
 @lersek
     tianocore member
lersek commented 16 days ago
Most recent kvm-arm discussion: http://thread.gmane.org/gmane.comp.emulators.kvm.arm.devel/7113
 @lerseklersek referenced this issue 16 days ago
 Open
RFE: OvmfPkg: GOP driver for virtio-gpu-pci #103
@lersek
     tianocore member
lersek commented 16 days ago
See also #103

**Comment by tianocore-issues** [comment]
## Comment 374
Date: 2016-07-21 17:12:26 +0000
From: @lersek

Confirming.

**Comment by tianocore-issues** [comment]
## Comment 447
Date: 2016-08-23 23:31:08 +0000
From: @lersek

Series from Ard:
https://lists.01.org/pipermail/edk2-devel/2016-August/000959.html

**Comment by tianocore-issues** [comment]
## Comment 462
Date: 2016-09-01 20:47:44 +0000
From: @lersek

v2 series from Ard:
https://lists.01.org/pipermail/edk2-devel/2016-August/001193.html

**Comment by tianocore-issues** [comment]
## Comment 463
Date: 2016-09-02 16:38:29 +0000
From: @lersek

v3 series from Ard:
https://lists.01.org/pipermail/edk2-devel/2016-September/001323.html

**Comment by tianocore-issues** [comment]
## Comment 464
Date: 2016-09-02 17:16:37 +0000
From: @lersek

Fixed by commit range 0b09c212a8ae..646a9e5b799b.
